GRPC CAS implementation

cachemeifyoucan opened the pull request:
A simple GRPC CAS prototype that uses sync APIs. The remote CAS is
backed up by a local in memory hash map trie to manage Ref/Hash/Object
lifetime.

For this CAS, ObjectRef is pointing to the index of the hash map trie
instead of the data. This allows the load only happens when you call
`load` on ObjectRef for an ObjectHandle (which still points to data).

cachemeifyoucan committed:
GRPC CAS Prototype

A simple GRPC CAS prototype that uses sync APIs. The remote CAS is
backed up by a local in memory hash map trie to manage Ref/Hash/Object
lifetime.

For this CAS, ObjectRef is pointing to the index of the hash map trie
instead of the data. This allows the load only happens when you call
`load` on ObjectRef for an ObjectHandle (which still points to data).

cachemeifyoucan commented on llvm/cmake/modules/FindGRPC.cmake:
```diff
@@ -132,7 +132,7 @@ function(generate_protos LibraryName ProtoFile)
         DEPENDS "${ProtoSourceAbsolutePath}")
 
-  add_clang_library(${LibraryName} ${GeneratedProtoSource}
+  add_llvm_library(${LibraryName} ${GeneratedProtoSource}
```
This can probably go upstream. We should not use add_clang_library in a llvm cmake module and it doesn't seem to break clangd build but I could be wrong.

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

benlangmuir commented on llvm/include/llvm-c/cas-plugin.h:
```diff
@@ -0,0 +1,38 @@
+|*                                                                            *|
+|* This header provides public interface to an abstract link time optimization*|
+|* library.  LLVM provides an implementation of this interface for use with   *|
+|* llvm bitcode files.                                                        *|
```
Header comment is from a different file 😺

cachemeifyoucan replied:
Good catch.

benlangmuir commented on llvm/include/llvm-c/cas-plugin.h:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Create a plugin CAS with arguments. Return NULL if an error occers.
+ */
+extern casdb_t llvm_create_plugin_cas(int argc, const char *argv[]);
```
Seems like we should specify in the documentation that it's the cas plugin that provides this API, not llvm itself.

benlangmuir commented on llvm/include/llvm/CAS/CASDB.h:
```diff
@@ -391,6 +391,9 @@ void getDefaultOnDiskCASStableID(SmallVectorImpl<char> &Path);
 std::string getDefaultOnDiskCASStableID();
 
+/// Create CASDB from a string identifier.
+Expected<std::unique_ptr<CASDB>> createCASFromIdentifier(StringRef Path);
```
We should document the URI schemes this supports.

cachemeifyoucan replied:
This is just a quick hack to integrate any CAS in with current options. Any other suggestion is welcomed.

benlangmuir commented on llvm/lib/CAS/CASDB.cpp:
```diff
@@ -138,3 +138,21 @@ CASDB::loadIndependentDataBuffer(ObjectHandle Node, const Twine &Name,
+    return createPluginCAS(Path);
+
+  if (Path.consume_back("file://"))
+    return createOnDiskCAS(Path);
```
If this is an actual URL we probably need to handle percent encoding (e.g. %20 for space).  Same for `plugin://` above.  I'm fine with just adding a FIXME comment for now.

cachemeifyoucan replied:
Ok. If we stick with URL schema, we can pass plugin arguments as:
`plugin://address#arguments` but we need to support the URL correctly for that.

benlangmuir reviewed (commented):
I only reviewed the core CAS changes, not the gRPC implementation.

cachemeifyoucan commented:
There is one major problem in this implementation is that llvm custom RTTI doesn't really work across dylib boundary. You can't correctly handle errors returned from Plugin. `clang` does not export necessary symbols for dyld to coalesce the base class so `isa<>` check will fail. It happens to work for release build for the cases we work but I would not upstream that for now.

Maybe the correct way to do is to create C API for GRPC and contain all the error handling on the clang side?

vlm commented on llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,638 @@
+  Address += Path;
+  ChannelArguments Args;
+  Args.SetMaxReceiveMessageSize(INT_MAX);
+  Args.SetMaxReceiveMessageSize(INT_MAX);
```
nit: Dupe?

vlm commented on llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,638 @@
+  std::string ret;
+
+  auto is_base64 = [](unsigned char c) {
+    return (isalnum(c) || (c == '+') || (c == '/'));
```
* `isalnum` can be very expensive in non-C locales. Maybe a custom range check would be better?
* If we're doing a custom decoding anyway, maybe switching to Base64URL would provide broader compatibility? https://www.rfc-editor.org/rfc/rfc4648#section-5

cachemeifyoucan replied:
The fixup above says this function (encoding and decoding of the hash) should be provided by CASService (through gRPC).

I implement the our hash encoding here just so that it is easier to debug. Technically, I can just hexdump the hash bytes and it would be the same.

vlm commented on llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,638 @@
+  if (!Reference.consume_front(getCASIDPrefix()))
+    return createStringError(std::make_error_code(std::errc::invalid_argument),
+                             "invalid cas-id '" + Reference + "'");
+  std::string Binary = decodeBase64(Reference.str());
```
May I recommend base64URL for identifiers?

vlm commented on llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,638 @@
+void GRPCRelayCAS::printIDImpl(raw_ostream &OS, const CASID &ID) const {
+  assert(&ID.getContext() == this);
+  ArrayRef<uint8_t> Hash = ID.getHash();
+  assert(Hash[0] == 0 && "Expected to start with 0");
```
I believe this constraint will be found rather inconvenient at some point, as the schemas evolve.

vlm commented on llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,638 @@
+#ifdef __APPLE__
+  // FIXME: Move to FileSystem.h?
+  // Get file path from FD.
+  char Path[1024];
```
nit: `s/1024/PATH_MAX/`?

cachemeifyoucan replied:
Sure, PATH_MAX is apple platform macro and do not exist on Windows. Since this is under `__APPLE__` PATH_MAX is fine I guess.

vlm commented on llvm/tools/libRemoteCachingService/RemoteCachingService.cpp:
```diff
@@ -0,0 +1,34 @@
+
+casdb_t llvm_create_plugin_cas(int argc, const char *argv[]) {
+  // If no argument was passed, use default path. Ignore other args.
+  std::string Path = argc == 0 ? "/tmp/compilation-cache-daemon" : argv[1];
```
nit: Using `confstr(_CS_DARWIN_USER_TEMP_DIR)` instead of `/tmp` often provides better security.

cachemeifyoucan replied:
Haha, no. Socket path length is just above 100 char and your system USER_TEMP_DIR can be longer than that (which is the case for our CI that decide to put it under the jenkins build directory).

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@benlangmuir @akyrtzi Rebase the GRPC CAS on top of current implementation.

cachemeifyoucan committed:
[RemoteCache] Unify two copies of remote cache grpc protocol

Unify the two copies of GRPC protocol in the code base after rebasing.

cachemeifyoucan committed:
[RemoteCache] Implement full GRPC CAS on top of async API

Add implementation for GRPC Put/Get API to enable use/test GRPCRelayCAS
on top of the async API. The API for the CAS is still sync'ed.

cachemeifyoucan committed:
[RemoteCachingService] Make building with or without RemoteCache cleaner

Sink the decision whether Remote Cache is enabled or not into
RemoteCachingService library so the user in general doesn't need to
worry about build dependency when use it.

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented on llvm/cmake/modules/LLVMConfig.cmake.in:
```diff
@@ -52,6 +52,7 @@ set(LLVM_ENABLE_ASSERTIONS @LLVM_ENABLE_ASSERTIONS@)
 set(LLVM_ENABLE_EH @LLVM_ENABLE_EH@)
 
 set(LLVM_ENABLE_ONDISK_CAS @LLVM_ENABLE_ONDISK_CAS@)
+set(LLVM_CAS_ENABLE_REMOTE_CACHE @LLVM_CAS_ENABLE_REMOTE_CACHE@)
```
Is there a way to avoid changing `llvm-config.h` based on whether `LLVM_CAS_ENABLE_REMOTE_CACHE` is enabled or not?

This header is affecting practically *every* compilation from LLVM, while whether remote cache is enabled or not should technically only affect a handful of places.

cachemeifyoucan replied:
But that is a build time configuration for whether the compile supports remote caching or not, not if remote caching is available or not. Are you expecting us to support both configuration in CAS CI?

The most useful place for this condition is actually unittest where we spin up different kinds of CAS to run on the same set of tests. I can change it to be based on runtime behavior but that will make breaking GRPC CAS undetectable.

akyrtzi replied:
I mean you don't need to put a macro in `llvm-config.h`; if you have C++ code that needs to check how `LLVM_CAS_ENABLE_REMOTE_CACHE` was configured then you could pass a macro via the `CMakeLists.txt` of the target that contains such code.

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented on llvm/include/llvm/CAS/ObjectStore.h:
```diff
@@ -367,6 +365,27 @@ void getDefaultOnDiskCASStableID(SmallVectorImpl<char> &Path);
+///  * InMemory CAS: mem://
+///  * OnDisk CAS: file://${PATH_TO_ONDISK_CAS}
+///  * PlugIn CAS: plugin://${PATH_TO_PLUGIN}
+/// If no URL scheme is used, it is default to following (but might change in
```
"it is default to" -> "it defaults to"

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -11,15 +11,18 @@
-namespace clang {
-namespace remote_cache {
+namespace llvm {
+namespace cas {
```
I think it would be worth it to wrap the types of this library in a `llvm::cas::remote` namespace; it'd be a bit clearer for code comprehension to see in code `llvm::cas::remote::ClientServices` vs `llvm::cas::ClientServices`.

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -70,17 +73,17 @@ class KeyValueDBClient {
       ++NumPending;
     }
-    void getValueAsync(ArrayRef<uint8_t> Key,
+    void getValueAsync(llvm::ArrayRef<uint8_t> Key,
```
The `llvm::` prefixes are not needed in this file.

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -221,25 +225,107 @@ class CASDBClient {
+    virtual ~GetAsyncQueue() = default;
+
+    void getAsync(std::string CASID,
+                  llvm::Optional<std::string> OutFilePath = llvm::None,
```
Potentially we could simplify this API a bit and have the blob data always coming as bytes from the channel, instead of providing the option for writing the blob to a file, but it's up to you (ditto for `PutAsyncQueue`, leave `putFileAsync` out?).

cachemeifyoucan replied:
I think we need those since the GRPC protocol support that. We also probably need to implementing returning something larger than 4MB in a file to workaround GRPC limit (unless we switch to streaming API).

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -221,25 +225,107 @@ class CASDBClient {
+    virtual llvm::Expected<Response> receiveNextImpl() = 0;
+  };
+
+  class GetAsyncQueue : public AsyncQueueBase {
```
You can remove the
```
FIXME: This only implements the \p CASSave and \p CASLoad APIs.
```
earlier in this file.

akyrtzi commented on llvm/include/llvm/RemoteCachingService/RemoteCacheServer.h:
```diff
@@ -1,49 +1,47 @@
-/// doing the actual work of storing and retrieving the data.
-RemoteCacheServer
-createServer(StringRef SocketPath,
-             std::unique_ptr<RemoteCacheProvider> CacheProvider);
```
I'd like to keep the flexibility of separation of concerns between `RemoteCacheServer` (handles the gRPC protocol channel) and `RemoteCacheProvider` (doing the work for retrieving/storing data).
That way it's very clear and straightforward how you would create an alternative implementation (e.g. one that connects to S3); you just write a new `RemoteCacheProvider` implementation.

cachemeifyoucan replied:
That doesn't work well because I want to hide all the grpc protocol from public headers (those in include/llvm), but in order to do anything to that `std::unique_ptr` inside the function signature, we need a concrete type for `RemoteCacheProvider` but that type has grpc dependency.

I decide that I will keep the header clean and we can figure out how to separate them cleanly when we have a second provider.

akyrtzi replied:
Ok. But note there is a leftover `class RemoteCacheProvider;` reference in `include/llvm/RemoteCachingService/RemoteCacheServer.h`.

akyrtzi commented on llvm/include/llvm/RemoteCachingService/RemoteCacheServer.h:
```diff
@@ -1,49 +1,47 @@
 
-  void Run();
+  void Start();
+  void Wait();
```
Is it really useful to separate `Run` to `Start` + `Wait`? `Wait` is a bit confusing, one could interpret it as "make the server pause".

cachemeifyoucan replied:
Yes, in order to setup the server in the same process as client, we need to have a setup function that returns when the channel is created on the server side, so the client can start to send message. We can keep the `Run()` by calling Start and Wait if needed.

See how it is used in MockServer in unittest.

akyrtzi replied:
Could we rename `Wait()` to `Listen()`?

akyrtzi commented on llvm/lib/CAS/ObjectStore.cpp:
```diff
@@ -135,3 +125,41 @@ ObjectProxy::getMemoryBuffer(StringRef Name,
+
+  // FIXME: some current default behavior.
+  if (Path.empty())
+    return createInMemoryCAS();
```
Is anything depending on this (getting in-memory CAS with empty string) apart from tests?
I feel that it'd be better for maintenance to treat passing an empty string as programming error and return an `llvm::Error`.

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,504 @@
+
+Expected<ObjectRef>
+GRPCRelayCAS::storeFromOpenFileImpl(sys::fs::file_t FD,
+                                    Optional<sys::fs::file_status> FS) {
```
What is this intended for, there's no caller in the PR. I'm wondering why doesn't this function accept a path directly? It seems you added `sys::fs::getRealPathFromHandle` just for this function.

cachemeifyoucan replied:
This is not a new function. It always exists and used by CAS file system and it has a different default implementation overwrite in BuiltinCAS and here is another one for GRPC CAS.

BuiltinCAS use this to store/use the underlying memory mapped file region directly and GRPC uses the original file directly to send to server.

akyrtzi replied:
Ok, I see it is supporting the `ObjectStore::storeFromOpenFile()` API.
It seems useful to add a `ObjectStore::storeFromFile(StringRef FilePath)`, so the caller doesn't go through hoops if they just have a file path (and the GRPC implementation just passes along the path), but this could be considered as follow-up.

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+set(LLVM_LINK_COMPONENTS
+  RemoteCacheKVProto
+  RemoteCacheCASProto
+  RemoteCacheClient
```
Can these 3 entries move after `LINK_COMPONENTS` below, inside `add_llvm_library()`?

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,504 @@
+  void anchor() override;
+
+public:
+  static StringRef getHashName() { return "Luxon"; }
```
How about "RemoteCAS" instead?

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,504 @@
+void GRPCCASContext::printIDImpl(raw_ostream &OS, const CASID &ID) const {
+  assert(&ID.getContext() == this);
+  SmallString<64> Hash;
+  toHex(ID.getHash(), /*LowerCase=*/true, Hash);
```
Recommend to use `llvm::encodeBase64()` for this.

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,504 @@
+
+Expected<CASID> GRPCRelayCAS::parseID(StringRef Reference) {
+  std::string Binary;
+  if (!tryGetFromHex(Reference, Binary))
```
Recommend to use `llvm::decodeBase64()`.

cachemeifyoucan replied:
Ah, decode is new!

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,504 @@
+  }
+
+  const InMemoryCASData &
+  storeObjectImpl(InMemoryIndexValueT &I,
```
I find it non-ideal that all the data are preserved in the process's memory, without the client having any control for the behavior (whether the data should just be relayed only or preserved in-memory as well, but we could iterate in a potential follow-up.

cachemeifyoucan replied:
The reason for this is simply because the life time of object is the same as CAS, so CAS needs to keep the memory alive till it is destroyed.

But yes, we can change the design but that is a much bigger problem than implement a GRPC CAS.

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,504 @@
+  return toReference(storeObjectImpl(I, None, Data));
+}
+
+Expected<ObjectRef> GRPCRelayCAS::storeFile(StringRef Content) {
```
I find `storeFile` confusing since it's not accepting a file path, it's accepting a data buffer.
How about `storeBlob` instead? And `loadBlob` below.

cachemeifyoucan replied:
This is currently dead code. I can remove it since it is to test Load and Save GRPC protocol, which is currently used. Maybe it should just be dropped if CAS is not using them.

akyrtzi replied:
I'm fine with dropping.

akyrtzi commented on llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp:
```diff
@@ -0,0 +1,504 @@
+
+  auto Result = Response->Value->find("CASID");
+  if (Result == Response->Value->end())
+    return None;
```
This seems like it should be an error, because it would be unexpected to get a result that didn't include "CASID", something went wrong.

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented on llvm/unittests/CAS/CMakeLists.txt:
```diff
@@ -1,7 +1,14 @@
+if (LLVM_CAS_ENABLE_REMOTE_CACHE)
+  append("-DLLVM_CAS_ENABLE_REMOTE_CACHE=1" CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
```
The idiomatic way to pass defines in CMake is as
```
add_definitions(-DLLVM_CAS_ENABLE_REMOTE_CACHE=1)
```

akyrtzi commented on llvm/unittests/CAS/ObjectStoreTest.cpp:
```diff
@@ -8,7 +8,9 @@
 
 #include "llvm/CAS/ObjectStore.h"
 #include "llvm/Config/llvm-config.h"
+#include "llvm/RemoteCachingService/RemoteCachingService.h"
```
This header doesn't seem to get used in this file.

akyrtzi commented on llvm/unittests/CAS/ObjectStoreTest.cpp:
```diff
@@ -8,7 +8,9 @@
 #include "llvm/Config/llvm-config.h"
+#include "llvm/RemoteCachingService/RemoteCachingService.h"
 #include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
```
Is this include needed?

akyrtzi commented on llvm/unittests/CAS/MockGRPCServer.cpp:
```diff
@@ -0,0 +1,70 @@
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/CAS/ObjectStore.h"
+#include "llvm/Config/llvm-config.h"
```
This include is probably not needed anymore.

cachemeifyoucan committed:
Address review feedback

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented on llvm/lib/RemoteCachingService/NullService/NullService.cpp:
```diff
@@ -0,0 +1,57 @@
+      "RemoteCachingService (LLVM_CAS_ENABLE_REMOTE_CACHE) is disabled");
+}
+
+Expected<std::unique_ptr<ObjectStore>> createGRPCRelayCAS(const Twine &Path) {
```
It's good to keep a prefix for the function definitions in this file (`remote::createGRPCRelayCAS(...)`), so that if the function signatures change you get a compiler error instead of a harder to diagnose linker error.

cachemeifyoucan commented:
@swift-ci please test
